compiler/rustc_typeck/src/check/upvar.rs

```diff
@@ -284,30 +284,17 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             let var_hir_id = upvar_id.var_path.hir_id;
             closure_captures.insert(var_hir_id, upvar_id);
 
-            let mut new_capture_kind = capture_info.capture_kind;
-            if let Some(existing_capture_kind) =
+            let new_capture_kind = if let Some(capture_kind) =
                 self.typeck_results.borrow_mut().upvar_capture_map.get(&upvar_id)
             {
-                // FIXME(@azhng): refactor this later
-                new_capture_kind = match (existing_capture_kind, new_capture_kind) {
-                    (ty::UpvarCapture::ByValue(Some(_)), _) => *existing_capture_kind,
-                    (_, ty::UpvarCapture::ByValue(Some(_))) => new_capture_kind,
-                    (ty::UpvarCapture::ByValue(_), _) | (_, ty::UpvarCapture::ByValue(_)) => {
-                        ty::UpvarCapture::ByValue(None)
-                    }
-                    (ty::UpvarCapture::ByRef(existing_ref), ty::UpvarCapture::ByRef(new_ref)) => {
-                        match (existing_ref.kind, new_ref.kind) {
-                            // Take RHS:
-                            (ty::ImmBorrow, ty::UniqueImmBorrow | ty::MutBorrow)
-                            | (ty::UniqueImmBorrow, ty::MutBorrow) => new_capture_kind,
-                            // Take LHS:
-                            (ty::ImmBorrow, ty::ImmBorrow)
-                            | (ty::UniqueImmBorrow, ty::ImmBorrow | ty::UniqueImmBorrow)
-                            | (ty::MutBorrow, _) => *existing_capture_kind,
-                        }
-                    }
-                };
-            }
+                // upvar_capture_map only stores the UpvarCapture (CaptureKind),
+                // so we create a fake capture info with no expression.
+                let fake_capture_info =
```

Comment:
fake capture info? I am kind of confused. So we choose between two types of different capture based on `determine_capture_info`, so is it really a fake_capture_info? Do you mean it's the old capture info?

Reply:
The reason it's fake is because it's not the actual capture_info that resulted in the existing entry in the map, espcially note the None expr_id